# API Data Report
## Paul Keren
### 10/11/2024
The hypothesis that I am testing is if "Happy" by Pharell is the happiest song of all time measured by valence, danceability, and energy. My theoritical application is that the song "Happy" is described as a very happy song in the title and in the lyrics of "Because I'm happy". Though there are many other happy songs that exist in the world so I will be testing the valence, danceability, and energy of each of these tracks to see which one measues the highest. 

The Spotify endpoints I will be using to collect my data are the tracks, features, and albums. The responses I will be using are the valence, energy, and danceability of a track, these respones are suitable for this project because valence measures the musical positiveness in a track so it is a good endpoint to use to test which of the tracks is the most happy. The other respones of danceability and energy are suitable for this hypothesis because if a track is very danceable it means it is upbeat and dancing makes people very happy to be in the moment and having a good time. For energy if a track has a lot of energy it makes the track very optimistic. Some limitations of the data can be user reaction, so we do not know totally how users would react to the top of this list. 

In [33]:
import urllib
import requests
import pandas as pd
import json
import base64

In [34]:
def get_session_token(SessionID, SessionKey):
    url = 'https://accounts.spotify.com/api/token'
    data = {'grant_type':'client_credentials'}
    encoded_key = base64.b64encode(str(SessionID + ":" + SessionKey).\
                                   encode("ascii"))
    header = {'Authorization': 'Basic {}'.format(encoded_key.decode("ascii"))}
    
    response = requests.post(url, 
                            data = data,
                           headers = header)
    print(response.status_code)
    return response.json()['access_token']

In [35]:
keys = pd.read_csv("Keys.txt")

In [36]:
access_token = get_session_token(keys['Client_ID'].iloc[0],keys['Client_Secret'].iloc[0])

200


The access token ^

In [37]:
access_token

'BQDFaQKO1XilzPxs_1z_OkgvFtMTr-B80ZAjKOUQ7sX8UVsPLbask1Y-TcWlagxpA_0s8adKynd_w7pyqBe3zCuBrsnPigz7bH6hsKFkkWocBoM7rF0'

In [38]:
t_features_ep = 'https://api.spotify.com/v1/audio-features/'
tracks_ep = 'https://api.spotify.com/v1/tracks'
ab_tracks_ep = 'https://api.spotify.com/v1/albums/{}/tracks'

The endpoints that are used ^

In [39]:
def api_call(endpoint_url, api_header):
    response = requests.get(endpoint_url, headers = api_header)
    print(response.status_code)
    return response.json()

In [40]:
session_header = {'Authorization': 'Bearer {}'.format(access_token)}

In [41]:
hy_album_id = '0jVGdTKQ0pValbbtxv5h5O'
hy2_album_id = '1FVsTVNu2YGVYcOjY7VtHk'

The albums used are mostly just compilations of generally happy songs or songs that are rated as happy songs, so the variables reflect that being hy = happy, and hy2 = happy 2

In [42]:
ab_tracks_ep.format(hy_album_id)

'https://api.spotify.com/v1/albums/0jVGdTKQ0pValbbtxv5h5O/tracks'

In [43]:
hy_album_response = api_call(ab_tracks_ep.format(hy_album_id) , session_header)
hy2_album_response = api_call(ab_tracks_ep.format(hy2_album_id) , session_header)

200
200


In [44]:
hy_album_df = pd.DataFrame(hy_album_response['items'])
hy2_album_df = pd.DataFrame(hy2_album_response['items']) 
#hy_album_df.head()

In [45]:
hy_track_ids = ','.join(hy_album_df['id'].to_list())
hy2_track_ids = ','.join(hy2_album_df['id'].to_list())

In [46]:
t_features_ep + '?ids={}'.format(hy_track_ids)

'https://api.spotify.com/v1/audio-features/?ids=1huwecVUu6sWgU7o7GHPHQ,2IoU47UbLPu8LX3eAnK7mj,6ZPg2irmu84pBiLUPBSR87,0Ad6w4dXEF0OI3HnZQNuuf,5GtQkJTQ01zxZ9xRuIBRyY,2b80sCsAwCpdWWMdcQYf7s,7fsLtpbscBZUQntENqmR1S,2B9rafHNGPpnaAtqTcbrFL,62tQ11UnK9za7j0dyqT7Hs,6ODJxBvPLPRgKUVFOHClqb,29TUsmgNleSNV7pDAuMpiD,2gEkq0Mrrd6sKUkBnxWqDq,0bP2WhCT0F2pA0vjqaiDSk,0cDbVbHEUM419vxXxigLtS,3IL9TPElS6uk1tgiKEUVie,6TFr1fKugaFyGF7hzH2XQo,6dzH7tWblVkIVRVdFL7sS4,2SMH7dslMqZLscuNIaIsLI,4WPcZxkaJJxAZcbuIVWbhK,51c0YDa7BZUiT79IwDZVkj'

In [47]:
hy_track_features = api_call(t_features_ep + '?ids={}'.format(hy_track_ids), 
                             session_header)
hy2_track_features = api_call(t_features_ep + '?ids={}'.format(hy2_track_ids), 
                             session_header)

200
200


In [48]:
hy_track_info = api_call(tracks_ep + '?market=US&ids={}'.format(hy_track_ids), 
                             session_header)
hy2_track_info = api_call(tracks_ep + '?market=US&ids={}'.format(hy2_track_ids), 
                             session_header)

200
200


In [49]:
hy_features_df = pd.DataFrame(hy_track_features['audio_features'])
hy2_features_df = pd.DataFrame(hy2_track_features['audio_features'])
hy_features_df.head()

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.647   0.823    5    -4.662     0       0.1840       0.22000   
1         0.853   0.681    0    -5.003     1       0.0888       0.00966   
2         0.652   0.749    0    -5.248     1       0.0886       0.08480   
3         0.637   0.876    5    -5.269     1       0.0366       0.07240   
4         0.595   0.858    4    -3.555     0       0.0345       0.02110   

   instrumentalness  liveness  valence    tempo            type  \
0          0.000000    0.0908    0.962  160.023  audio_features   
1          0.000057    0.0233    0.930  114.989  audio_features   
2          0.000000    0.6040    0.651   94.103  audio_features   
3          0.000000    0.2150    0.433  116.943  audio_features   
4          0.017500    0.1410    0.737  127.949  audio_features   

                       id                                   uri  \
0  1huwecVUu6sWgU7o7GHPHQ  spotify:track:1huwecVUu6sWgU7o7GHPHQ   
1  2IoU47UbLPu8LX3eAnK7mj  spotify:track:2IoU47UbLPu8LX3eAnK7mj   
2  6ZPg2irmu84pBiLUPBSR87  spotify:track:6ZPg2irmu84pBiLUPBSR87   
3  0Ad6w4dXEF0OI3HnZQNuuf  spotify:track:0Ad6w4dXEF0OI3HnZQNuuf   
4  5GtQkJTQ01zxZ9xRuIBRyY  spotify:track:5GtQkJTQ01zxZ9xRuIBRyY   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/1huwecVUu6sW...   
1  https://api.spotify.com/v1/tracks/2IoU47UbLPu8...   
2  https://api.spotify.com/v1/tracks/6ZPg2irmu84p...   
3  https://api.spotify.com/v1/tracks/0Ad6w4dXEF0O...   
4  https://api.spotify.com/v1/tracks/5GtQkJTQ01zx...   

                                        analysis_url  duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/1huw...       232720   
1  https://api.spotify.com/v1/audio-analysis/2IoU...       270627   
2  https://api.spotify.com/v1/audio-analysis/6ZPg...       267413   
3  https://api.spotify.com/v1/audio-analysis/0Ad6...       235733   
4  https://api.spotify.com/v1/audio-analysis/5GtQ...       222533   

   time_signature  
0               4  
1               4  
2               4  
3               4  
4               4

In [50]:
hy_tracks_df = pd.DataFrame(hy_track_info['tracks'])
hy2_tracks_df = pd.DataFrame(hy2_track_info['tracks'])
hy_tracks_df.head()


album  \
0  {'album_type': 'compilation', 'artists': [{'ex...   
1  {'album_type': 'compilation', 'artists': [{'ex...   
2  {'album_type': 'compilation', 'artists': [{'ex...   
3  {'album_type': 'compilation', 'artists': [{'ex...   
4  {'album_type': 'compilation', 'artists': [{'ex...   

                                             artists  disc_number  \
0  [{'external_urls': {'spotify': 'https://open.s...            1   
1  [{'external_urls': {'spotify': 'https://open.s...            1   
2  [{'external_urls': {'spotify': 'https://open.s...            1   
3  [{'external_urls': {'spotify': 'https://open.s...            1   
4  [{'external_urls': {'spotify': 'https://open.s...            1   

   duration_ms  explicit              external_ids  \
0       232720     False  {'isrc': 'USQ4E1300686'}   
1       270626      True  {'isrc': 'GBARL1401524'}   
2       267413     False  {'isrc': 'USSM11102908'}   
3       235733     False  {'isrc': 'USSM11401744'}   
4       222533     False  {'isrc': 'GBARL1400296'}   

                                       external_urls  \
0  {'spotify': 'https://open.spotify.com/track/1h...   
1  {'spotify': 'https://open.spotify.com/track/2I...   
2  {'spotify': 'https://open.spotify.com/track/6Z...   
3  {'spotify': 'https://open.spotify.com/track/0A...   
4  {'spotify': 'https://open.spotify.com/track/5G...   

                                                href                      id  \
0  https://api.spotify.com/v1/tracks/1huwecVUu6sW...  1huwecVUu6sWgU7o7GHPHQ   
1  https://api.spotify.com/v1/tracks/2IoU47UbLPu8...  2IoU47UbLPu8LX3eAnK7mj   
2  https://api.spotify.com/v1/tracks/6ZPg2irmu84p...  6ZPg2irmu84pBiLUPBSR87   
3  https://api.spotify.com/v1/tracks/0Ad6w4dXEF0O...  0Ad6w4dXEF0OI3HnZQNuuf   
4  https://api.spotify.com/v1/tracks/5GtQkJTQ01zx...  5GtQkJTQ01zxZ9xRuIBRyY   

   is_local  is_playable                            name  popularity  \
0     False         True  Happy - From "Despicable Me 2"          28   
1     False         True  Uptown Funk (feat. Bruno Mars)          28   
2     False         True                     Love On Top          23   
3     False         True         Love Never Felt So Good          20   
4     False         True                          Summer          32   

                                         preview_url  track_number   type  \
0  https://p.scdn.co/mp3-preview/7df50d6478d9366d...             1  track   
1  https://p.scdn.co/mp3-preview/6b518c4b49482770...             2  track   
2  https://p.scdn.co/mp3-preview/b33aaf728bc4a935...             3  track   
3  https://p.scdn.co/mp3-preview/84840c0327e036ed...             4  track   
4  https://p.scdn.co/mp3-preview/f8877a2ec4963cb1...             5  track   

                                    uri linked_from  
0  spotify:track:1huwecVUu6sWgU7o7GHPHQ         NaN  
1  spotify:track:2IoU47UbLPu8LX3eAnK7mj         NaN  
2  spotify:track:6ZPg2irmu84pBiLUPBSR87         NaN  
3  spotify:track:0Ad6w4dXEF0OI3HnZQNuuf         NaN  
4  spotify:track:5GtQkJTQ01zxZ9xRuIBRyY         NaN

We start getting good data here but we need to wrap it up more

In [51]:
hy_merged = pd.merge(hy_features_df, hy_tracks_df,
         how = 'inner', on = 'id')
hy_merged.head()

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.647   0.823    5    -4.662     0       0.1840       0.22000   
1         0.853   0.681    0    -5.003     1       0.0888       0.00966   
2         0.652   0.749    0    -5.248     1       0.0886       0.08480   
3         0.637   0.876    5    -5.269     1       0.0366       0.07240   
4         0.595   0.858    4    -3.555     0       0.0345       0.02110   

   instrumentalness  liveness  valence  ...  \
0          0.000000    0.0908    0.962  ...   
1          0.000057    0.0233    0.930  ...   
2          0.000000    0.6040    0.651  ...   
3          0.000000    0.2150    0.433  ...   
4          0.017500    0.1410    0.737  ...   

                                                href is_local is_playable  \
0  https://api.spotify.com/v1/tracks/1huwecVUu6sW...    False        True   
1  https://api.spotify.com/v1/tracks/2IoU47UbLPu8...    False        True   
2  https://api.spotify.com/v1/tracks/6ZPg2irmu84p...    False        True   
3  https://api.spotify.com/v1/tracks/0Ad6w4dXEF0O...    False        True   
4  https://api.spotify.com/v1/tracks/5GtQkJTQ01zx...    False        True   

                             name popularity  \
0  Happy - From "Despicable Me 2"         28   
1  Uptown Funk (feat. Bruno Mars)         28   
2                     Love On Top         23   
3         Love Never Felt So Good         20   
4                          Summer         32   

                                         preview_url  track_number  type_y  \
0  https://p.scdn.co/mp3-preview/7df50d6478d9366d...             1   track   
1  https://p.scdn.co/mp3-preview/6b518c4b49482770...             2   track   
2  https://p.scdn.co/mp3-preview/b33aaf728bc4a935...             3   track   
3  https://p.scdn.co/mp3-preview/84840c0327e036ed...             4   track   
4  https://p.scdn.co/mp3-preview/f8877a2ec4963cb1...             5   track   

                                  uri_y linked_from  
0  spotify:track:1huwecVUu6sWgU7o7GHPHQ         NaN  
1  spotify:track:2IoU47UbLPu8LX3eAnK7mj         NaN  
2  spotify:track:6ZPg2irmu84pBiLUPBSR87         NaN  
3  spotify:track:0Ad6w4dXEF0OI3HnZQNuuf         NaN  
4  spotify:track:5GtQkJTQ01zxZ9xRuIBRyY         NaN  

[5 rows x 35 columns]

More good data as we can see the variables we are going to measure

In [52]:
hy_merged.columns

Index(['danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness',
       'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo',
       'type_x', 'id', 'uri_x', 'track_href', 'analysis_url', 'duration_ms_x',
       'time_signature', 'album', 'artists', 'disc_number', 'duration_ms_y',
       'explicit', 'external_ids', 'external_urls', 'href', 'is_local',
       'is_playable', 'name', 'popularity', 'preview_url', 'track_number',
       'type_y', 'uri_y', 'linked_from'],
      dtype='object')

In [53]:
hy2_merged = pd.merge(hy2_features_df, hy2_tracks_df,
                     how = 'inner', on = 'id')
hy_merged['artist'] = "Various Artists"
hy2_merged['artist'] = "Multiple Artists"
hy2_merged.shape

(20, 34)

Again the albums are compilations so there is no specific artist for each song, hence why the artist name is Various and Multiple

In [54]:
artist_compare = pd.concat([hy_merged, hy2_merged], axis = 0)
artist_compare

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.647   0.823    5    -4.662     0       0.1840      0.220000   
1          0.853   0.681    0    -5.003     1       0.0888      0.009660   
2          0.652   0.749    0    -5.248     1       0.0886      0.084800   
3          0.637   0.876    5    -5.269     1       0.0366      0.072400   
4          0.595   0.858    4    -3.555     0       0.0345      0.021100   
5          0.606   0.720    5    -6.253     0       0.0283      0.000149   
6          0.654   0.868    1    -3.053     1       0.0458      0.025600   
7          0.811   0.879    9    -3.724     1       0.0513      0.047600   
8          0.700   0.747    5    -3.453     1       0.0964      0.274000   
9          0.827   0.580    4    -7.793     1       0.0374      0.179000   
10         0.699   0.870    0    -4.925     1       0.0546      0.000532   
11         0.500   0.788    6    -3.713     1       0.0484      0.012000   
12         0.604   0.631    1    -5.841     0       0.0360      0.008410   
13         0.750   0.714    2    -4.646     0       0.0957      0.063600   
14         0.758   0.878   11    -6.398     0       0.1280      0.006040   
15         0.581   0.817    0    -6.190     0       0.0689      0.000527   
16         0.450   0.951    0    -4.160     0       0.1370      0.010300   
17         0.457   0.891    9    -7.004     1       0.1060      0.002670   
18         0.490   0.885    3    -4.121     1       0.0396      0.002180   
0          0.537   0.882    0    -2.306     1       0.0403      0.000350   
1          0.749   0.748    0    -5.509     1       0.0447      0.276000   
2          0.611   0.583    1   -14.182     1       0.0668      0.057000   
3          0.521   0.815    0    -6.148     1       0.0549      0.518000   
4          0.804   0.552    0    -7.109     1       0.1130      0.400000   
5          0.603   0.522    7    -7.600     1       0.0978      0.335000   
6          0.649   0.727    1    -4.865     1       0.0393      0.001540   
7          0.762   0.948    8    -5.151     1       0.1010      0.186000   
8          0.915   0.591    0    -8.865     1       0.0597      0.000468   
9          0.724   0.810    9    -8.532     0       0.0404      0.556000   
10         0.392   0.629    5    -7.056     1       0.0334      0.718000   
11         0.685   0.934    9    -7.104     1       0.0460      0.004500   
12         0.857   0.796    9    -6.866     0       0.0527      0.079000   
13         0.581   0.880    7    -4.390     1       0.0289      0.001050   
14         0.414   0.950    4    -4.676     1       0.0622      0.001240   
15         0.793   0.650    6    -8.298     0       0.0433      0.228000   
16         0.625   0.956    7    -4.730     1       0.0299      0.283000   
17         0.542   0.606    5    -5.817     1       0.0394      0.368000   
18         0.682   0.846    1    -6.345     0       0.2180      0.216000   
19         0.544   0.875   10    -5.129     0       0.0798      0.024100   

    instrumentalness  liveness  valence  ...  is_local is_playable  \
0           0.000000    0.0908    0.962  ...     False        True   
1           0.000057    0.0233    0.930  ...     False        True   
2           0.000000    0.6040    0.651  ...     False        True   
3           0.000000    0.2150    0.433  ...     False        True   
4           0.017500    0.1410    0.737  ...     False        True   
5           0.001650    0.0745    0.178  ...     False        True   
6           0.000000    0.0766    0.542  ...     False        True   
7           0.000002    0.1090    0.962  ...     False        True   
8           0.000000    0.4860    0.683  ...     False        True   
9           0.030100    0.0699    0.381  ...     False        True   
10          0.021900    0.3590    0.778  ...     False        True   
11          0.000000    0.0646    0.433  ...     False        True   
12          0.000000    0.3980    0.511  ...     Fals

In [55]:
name_valence_energy_danceability = hy_merged[["name", "valence", "energy", "danceability"]]

In [56]:
name_valence_energy_danceability.head(30)

name  valence  energy  \
0                      Happy - From "Despicable Me 2"    0.962   0.823   
1                      Uptown Funk (feat. Bruno Mars)    0.930   0.681   
2                                         Love On Top    0.651   0.749   
3                             Love Never Felt So Good    0.433   0.876   
4                                              Summer    0.737   0.858   
5   All of Me - Tiësto's Birthday Treatment Remix ...    0.178   0.720   
6                                      Best Song Ever    0.542   0.868   
7                                 All About That Bass    0.962   0.879   
8                                               Bills    0.683   0.747   
9        Cheerleader (Felix Jaehn Remix) - Radio Edit    0.381   0.580   
10                                    Clap Your Hands    0.778   0.870   
11                                     Heartbeat Song    0.433   0.788   
12                                      We Can't Stop    0.511   0.631   
13                                            TiK ToK    0.665   0.714   
14  Waka Waka (This Time for Africa) (feat. Freshl...    0.719   0.878   
15                    Party All Night (Sleep All Day)    0.800   0.817   
16                                        Rock N Roll    0.729   0.951   
17                                 I Wanna Get Better    0.384   0.891   
18                                        Superheroes    0.640   0.885   

    danceability  
0          0.647  
1          0.853  
2          0.652  
3          0.637  
4          0.595  
5          0.606  
6          0.654  
7          0.811  
8          0.700  
9          0.827  
10         0.699  
11         0.500  
12         0.604  
13         0.750  
14         0.758  
15         0.581  
16         0.450  
17         0.457  
18         0.490

Using our responses that we are highlighting for the hypothesis and putting them in a list of their own

In [57]:
name_valence_energy_danceability = hy2_merged[["name", "valence", "energy", "danceability"]]

In [58]:
name_valence_energy_danceability.head(30)

name  valence  energy  \
0                                  Here It Goes Again    0.810   0.882   
1                                         How Bizarre    0.797   0.748   
2                                        Superstition    0.877   0.583   
3                                    It's Not Unusual    0.950   0.815   
4                                        5 Years Time    0.749   0.552   
5                             I Got You (I Feel Good)    0.705   0.522   
6                                     Come And Get It    0.776   0.727   
7                                      Take Your Mama    0.937   0.948   
8                          Funkytown - Single Version    0.315   0.591   
9                     I Will Survive - Single Version    0.738   0.810   
10                                Wouldn't It Be Nice    0.720   0.629   
11                            MMMBop - Single Version    0.680   0.934   
12                                        Super Freak    0.944   0.796   
13                                     There She Goes    0.945   0.880   
14                                    T-Shirt Weather    0.552   0.950   
15                                        Upside Down    0.647   0.650   
16                                            Shuffle    0.389   0.956   
17                                          Baby Love    0.491   0.606   
18  (We're Gonna) Rock Around The Clock - Single V...    0.763   0.846   
19                                             Ooh La    0.782   0.875   

    danceability  
0          0.537  
1          0.749  
2          0.611  
3          0.521  
4          0.804  
5          0.603  
6          0.649  
7          0.762  
8          0.915  
9          0.724  
10         0.392  
11         0.685  
12         0.857  
13         0.581  
14         0.414  
15         0.793  
16         0.625  
17         0.542  
18         0.682  
19         0.544

The conclusion we can draw from this data is that "All About that Bass" by Meghan Trainor is regarded as the happiest song based on the average of valence, energy, and danceability. "All About that Bass" scored an average of 0.884, while coming in second was "Take Your Mama" with an average of 0.882. The top five rolls out with "Super Freak" by Rick James with a score of 0.866, "Uptown Funk" by Mark Ronson and Bruno Mars at fourth with an average of 0.821 and finally, the song I hypothesized would be the happiest song, "Happy" by Pharrell came in fifth with a score of 0.811. 

Some limitations of this data can be that user reaction is not shown in these charts as this is just data from Spotify itself on the dimensions of the track and not people's actual feelings about a track. An alternative approache could be creating a poll or something for individuals to vote on for them to decide on what song makes them the happiest, a next step could be propsing to data to people and asking them if they agreed with my findings or they feel some songs should be moved around.  